runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -3664,6 +3664,151 @@ TR::Register *J9::X86::TreeEvaluator::longBitCount(TR::Node *node, TR::CodeGener
+   
+   // Object not array, inline checks
+   generateLabelInstruction(LABEL, node, notArrayLabel, cg);
+   
+   // Null test
+   if (node->getOpCodeValue() != TR::checkcastAndNULLCHK)
+      {
+      generateRegRegInstruction(TESTRegReg(), node, ObjReg, ObjReg, cg);
+      generateLabelInstruction(JE4, node, fallThruLabel, cg);
+      }
+   
+   // load object class in temp3
+   generateLoadJ9Class(node, temp3Reg, ObjReg, cg);
+   
+   // cast class is interface
+   generateMemImmInstruction(TEST4MemImm4, node,
+       generateX86MemoryReference(temp2Reg, offsetof(J9ROMClass, modifiers), cg), J9AccInterface, cg);
+   generateLabelInstruction(JE4, node, isClassLabel, cg);
+   
+   // Obtain I-Table
+   generateRegMemInstruction(LRegMem(), node, temp1Reg, generateX86MemoryReference(temp3Reg, offsetof(J9Class, iTable), cg), cg);
+   // Loop through I-Table
+   generateLabelInstruction(LABEL, node, iTableLoopLabel, cg);
+   generateRegRegInstruction(TESTRegReg(), node, temp1Reg, temp1Reg, cg);
+   generateLabelInstruction(JE4, node, throwLabel, cg);
+   auto interfaceMR = generateX86MemoryReference(temp1Reg, offsetof(J9ITable, interfaceClass), cg);
+   generateMemRegInstruction(CMPMemReg(), node, interfaceMR, castClassReg, cg);
+   generateRegMemInstruction(LRegMem(), node, temp1Reg, generateX86MemoryReference(temp1Reg, offsetof(J9ITable, next), cg), cg);
+   generateLabelInstruction(JNE4, node, iTableLoopLabel, cg);
+   
+   // Found from I-Table
+   generateLabelInstruction(JMP4, node, fallThruLabel, cg);
+   
+   // cast class is non-interface class
+   generateLabelInstruction(LABEL, node, isClassLabel, cg);
+   // equality test
+   generateRegRegInstruction(CMPRegReg(use64BitClasses), node, temp3Reg, castClassReg, cg);
+   generateLabelInstruction(JE4, node, fallThruLabel, cg);
+   
+   // class not equal,
+   // if cast class depth >= min depth, compare with object class depth
+   // store cast class depth in temp2 for later use
+   // class depth mask must be low 16 bits to safely load without the mask.
+   static_assert(J9AccClassDepthMask == 0xffff, "J9_JAVA_CLASS_DEPTH_MASK must be 0xffff");
+   generateRegMemInstruction(TR::Compiler->target.is64Bit()? MOVZXReg8Mem2 : MOVZXReg4Mem2, node,
+            temp2Reg, generateX86MemoryReference(castClassReg, offsetof(J9Class, classDepthAndFlags), cg), cg);
+   generateRegImmInstruction(CMP2RegImm2, node, temp2Reg, ((int16_t)(cg->comp()->getOptions()->_minimumSuperclassArraySize)), cg);
+   generateLabelInstruction(JB4, node, checkSuperClassArrayLabel, cg);
+   
+   // cast class depth >= obj class depth, fail
+   generateRegMemInstruction(CMP2RegMem, node, temp2Reg, generateX86MemoryReference(temp3Reg, offsetof(J9Class, classDepthAndFlags), cg), cg);
+   generateLabelInstruction(JAE4, node, throwLabel, cg);
+   
+   // check obj class's super class array entry
+   generateLabelInstruction(LABEL, node, checkSuperClassArrayLabel, cg);
```

Comment:
Just FYI, an alternative sequences requiring one less instruction may be:
```
SHL temp2Reg, 3 for 64-bit or 2 for 32-bit
ADD temp2Reg, [temp3Reg, superclasses offset]
CMP classClassReg, [temp2Reg]
```
I'm not sure which one is better. Your approach has one less instruction but uses one more register comparing to this alternative.

Reply:
Thanks for pointing it out. On 64 bit, the extra reg isn't likely to cause significant register pressure. On 32 bit, it could put more register pressure due to limited number of regs. Since 64-bit is more prevalent, I think we could optimize for 64bit and leave as is, and I added comments here to reflect this tradeoff made.